## Connect to database file

In [74]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [75]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Tables and views on database

In [76]:
%%sql

SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

Here is also the database schema 

![Here is also the database schema](https://s3.amazonaws.com/dq-content/191/chinook-schema.svg)

## Genres that sell most tracks in the US

In [77]:
%%sql

WITH tracks_sold_in_USA AS
     (
      SELECT il.* 
      FROM invoice_line il
      INNER JOIN invoice i ON i.invoice_id = il.invoice_id
      INNER JOIN customer c ON c.customer_id = i.customer_id
      WHERE c.country = "USA"
     )


SELECT
    g.name genre,
    count(tusa.invoice_line_id) tracks_sold,
    round(cast(count(tusa.invoice_line_id) AS FLOAT) / (
     SELECT COUNT(*) 
            FROM tracks_sold_in_USA)
          ,4)
     percentage_sold
FROM tracks_sold_in_usa tusa
INNER JOIN track t on t.track_id = tusa.track_id
INNER JOIN genre g on g.genre_id = t.genre_id
GROUP BY 1
ORDER BY 2 DESC;


Done.


[('Rock', 561, 0.5338),
 ('Alternative & Punk', 130, 0.1237),
 ('Metal', 124, 0.118),
 ('R&B/Soul', 53, 0.0504),
 ('Blues', 36, 0.0343),
 ('Alternative', 35, 0.0333),
 ('Latin', 22, 0.0209),
 ('Pop', 22, 0.0209),
 ('Hip Hop/Rap', 20, 0.019),
 ('Jazz', 14, 0.0133),
 ('Easy Listening', 13, 0.0124),
 ('Reggae', 6, 0.0057),
 ('Electronica/Dance', 5, 0.0048),
 ('Classical', 4, 0.0038),
 ('Heavy Metal', 3, 0.0029),
 ('Soundtrack', 2, 0.0019),
 ('TV Shows', 1, 0.001)]

The three arists are chosen based on the most sold genres in the US. 
 1st - Red Tone - Punk - Percentage of genre sold in US: 53.38%
 2nd - Slim Jim Bites - Blues - Percentage of genre sold in US: 3.43%
 3rd - Meteor and the Girls - Pop - Percentage of genre sold in US: 2.09%

### Total amount of sales by sales support agent

In [78]:
%%sql
SELECT 
e.employee_id ID,
e.first_name || " " || e.last_name as Name,
e.country as Country,
e.hire_date,
ROUND(sum(i.total),2) total_sales
FROM customer c
INNER JOIN invoice i ON i.customer_id = c.customer_id
INNER JOIN employee e ON e.employee_id = c.support_rep_id
group by 2;

Done.


[(3, 'Jane Peacock', 'Canada', '2017-04-01 00:00:00', 1731.51),
 (4, 'Margaret Park', 'Canada', '2017-05-03 00:00:00', 1584.0),
 (5, 'Steve Johnson', 'Canada', '2017-10-17 00:00:00', 1393.92)]

Looking only and the total_sales column, we can see that Jane is the agent with the biggest amount of sales. But the hiring date allows us to correctly interpret these results. 

While her and Margaret may have the same performance (Jane's sales 9.3% bigger than Margaret's with the latter entering 1 month later the company) Steve has proven to be the best performer. That is because even though he has 24% less of sales than Jane, he entered the company 6 months later! 

### Analyze Sales by Country

In [79]:
%%sql

WITH country_or_other AS
    (
     SELECT CASE 
           WHEN (SELECT COUNT(*) 
                        FROM customer
                        WHERE country = c.country) = 1 THEN "Other"
           ELSE c.country
        END AS country,
        c.customer_id,
        il.*
        FROM customer c 
        INNER JOIN invoice i ON i.customer_id = c.customer_id
        INNER JOIN invoice_line il ON il.invoice_id = i.invoice_id
     )

SELECT country,
       customers,
       total_sales,
       avg_sales_customer,
       avg_order_value
    FROM (SELECT country,
                 count(distinct customer_id) customers,
                 sum(unit_price) total_sales,
                 sum(unit_price)/count(customer_id) avg_sales_customer,
                 sum(unit_price)/count(invoice_id) avg_order_value,
                 CASE 
                     WHEN country = "Other" THEN 1
                     ELSE 0
                     END AS sort
          FROM country_or_other
    GROUP BY 1
    ORDER BY sort ASC, total_sales DESC
    );
    
                



                

Done.


[('USA', 13, 1040.490000000008, 0.9900000000000075, 0.9900000000000075),
 ('Canada', 8, 535.5900000000034, 0.9900000000000063, 0.9900000000000063),
 ('Brazil', 5, 427.68000000000245, 0.9900000000000057, 0.9900000000000057),
 ('France', 5, 389.0700000000021, 0.9900000000000053, 0.9900000000000053),
 ('Germany', 4, 334.6200000000016, 0.9900000000000048, 0.9900000000000048),
 ('Czech Republic', 2, 273.24000000000103, 0.9900000000000038, 0.9900000000000038),
 ('United Kingdom', 3, 245.52000000000078, 0.9900000000000031, 0.9900000000000031),
 ('Portugal', 2, 185.13000000000022, 0.9900000000000012, 0.9900000000000012),
 ('India', 2, 183.1500000000002, 0.9900000000000011, 0.9900000000000011),
 ('Other', 15, 1094.9400000000085, 0.9900000000000077, 0.9900000000000077)]

### Number and percentage of album purchases

In [80]:
%%sql 
WITH invoice_first_track AS
     (
      SELECT invoice_id,
             MIN(track_id) first_track_id
             FROM invoice_line
         GROUP BY 1
     )
    

SELECT album_purchase,
       COUNT(invoice_id) total_invoices,
       CAST(COUNT(invoice_id) AS FLOAT) / (SELECT COUNT(*)
                            FROM invoice) percent
        FROM
        (SELECT ift.*,
                CASE
                    WHEN (   SELECT t.track_id FROM track t
                             WHERE t.album_id = (SELECT t2.album_id FROM track t2
                                             WHERE t2.track_id = ift.first_track_id)
                         
          
                       EXCEPT 
                              SELECT il.track_id FROM invoice_line il
                              WHERE il.invoice_id = ift.invoice_id
                          ) IS NULL
                    AND
                        (      SELECT il.track_id FROM invoice_line il
                               WHERE il.invoice_id = ift.invoice_id 
                    
                         EXCEPT
                         
                            SELECT t.track_id FROM track t
                            WHERE t.album_id = (SELECT t2.album_id FROM track t2
                                             WHERE t2.track_id = ift.first_track_id)
                         ) IS NULL
                    THEN "Yes"
                    ELSE "No"
                END AS album_purchase
           FROM invoice_first_track ift)
         GROUP BY album_purchase

Done.


[('No', 500, 0.8143322475570033), ('Yes', 114, 0.18566775244299674)]

I wouldn't advice to remove full album purchases since full-album purchases represent almost 20% of invoice.

### Most used artist in playlists

In [81]:
%%sql

SELECT ar.name,
       COUNT(distinct pt.playlist_id) number_of_playlists,
       COUNT(distinct pt.track_id) number_of_tracks
FROM playlist_track pt
INNER JOIN track t ON t.track_id = pt.track_id
INNER JOIN album al ON al.album_id = t.album_id
INNER JOIN artist ar ON ar.artist_id = al.artist_id
GROUP BY 1
ORDER BY 2 DESC, 3 DESC


Done.


[('Eugene Ormandy', 7, 3),
 ('Berliner Philharmoniker & Herbert Von Karajan', 6, 3),
 ('Academy of St. Martin in the Fields & Sir Neville Marriner', 6, 2),
 ('English Concert & Trevor Pinnock', 6, 2),
 ("The King's Singers", 6, 2),
 ('Michael Tilson Thomas & San Francisco Symphony', 5, 2),
 ('Academy of St. Martin in the Fields, Sir Neville Marriner & Thurston Dart', 5, 1),
 ('Adrian Leaper & Doreen de Feis', 5, 1),
 ('Alberto Turco & Nova Schola Gregoriana', 5, 1),
 ('Antal Doráti & London Symphony Orchestra', 5, 1),
 ('Barry Wordsworth & BBC Concert Orchestra', 5, 1),
 ('Berliner Philharmoniker & Hans Rosbaud', 5, 1),
 ('Boston Symphony Orchestra & Seiji Ozawa', 5, 1),
 ('Britten Sinfonia, Ivor Bolton & Lesley Garrett', 5, 1),
 ('Chicago Symphony Chorus, Chicago Symphony Orchestra & Sir Georg Solti', 5, 1),
 ('Chicago Symphony Orchestra & Fritz Reiner', 5, 1),
 ('Chor der Wiener Staatsoper, Herbert Von Karajan & Wiener Philharmoniker', 5, 1),
 ('Emanuel Ax, Eugene Ormandy & Philadelphia Orchestra', 5, 1),
 ('Gustav Mahler', 5, 1),
 ('James Levine', 5, 1),
 ('Julian Bream', 5, 1),
 ("Kent Nagano and Orchestre de l'Opéra de Lyon", 5, 1),
 ('Luciano Pavarotti', 5, 1),
 ('Martin Roscoe', 5, 1),
 ('Maurizio Pollini', 5, 1),
 ('Mela Tenenbaum, Pro Musica Prague & Richard Kapp', 5, 1),
 ('Orchestre Révolutionnaire et Romantique & John Eliot Gardiner', 5, 1),
 ('Philharmonia Orchestra & Sir Neville Marriner', 5, 1),
 ('Philip Glass Ensemble', 5, 1),
 ('Richard Marlow & The Choir of Trinity College, Cambridge', 5, 1),
 ('Scholars Baroque Ensemble', 5, 1),
 ('Sir Georg Solti & Wiener Philharmoniker', 5, 1),
 ('Sir Georg Solti, Sumi Jo & Wiener Philharmoniker', 5, 1),
 ('Ton Koopman', 5, 1),
 ('Wilhelm Kempff', 5, 1),
 ('Yo-Yo Ma', 5, 1),
 ('Iron Maiden', 4, 213),
 ('Metallica', 4, 112),
 ('Pearl Jam', 4, 67),
 ('Eric Clapton', 4, 48),
 ('Chico Buarque', 4, 34),
 ('Gilberto Gil', 4, 32),
 ('Ozzy Osbourne', 4, 32),
 ('Nirvana', 4, 29),
 ('Djavan', 4, 26),
 ('Caetano Veloso', 4, 21),
 ('Mötley Crüe', 4, 17),
 ('Soundgarden', 4, 17),
 ('Gonzaguinha', 4, 14),
 ('Alice In Chains', 4, 12),
 ('Stone Temple Pilots', 4, 12),
 ('Temple of the Dog', 4, 10),
 ('Accept', 4, 4),
 ('Aaron Copland & London Symphony Orchestra', 4, 1),
 ('Academy of St. Martin in the Fields Chamber Ensemble & Sir Neville Marriner', 4, 1),
 ('Academy of St. Martin in the Fields, John Birch, Sir Neville Marriner & Sylvia McNair', 4, 1),
 ('Anne-Sophie Mutter, Herbert Von Karajan & Wiener Philharmoniker', 4, 1),
 ('Berliner Philharmoniker, Claudio Abbado & Sabine Meyer', 4, 1),
 ('C. Monteverdi, Nigel Rogers - Chiaroscuro; London Baroque; London Cornett & Sackbu', 4, 1),
 ("Charles Dutoit & L'Orchestre Symphonique de Montréal", 4, 1),
 ('Choir Of Westminster Abbey & Simon Preston', 4, 1),
 ("Christopher O'Riley", 4, 1),
 ('Edo de Waart & San Francisco Symphony', 4, 1),
 ('Emerson String Quartet', 4, 1),
 ('Equale Brass Ensemble, John Eliot Gardiner & Munich Monteverdi Orchestra and Choir', 4, 1),
 ('Felix Schmidt, London Symphony Orchestra & Rafael Frühbeck de Burgos', 4, 1),
 ('Fretwork', 4, 1),
 ('Gerald Moore', 4, 1),
 ('Göteborgs Symfoniker & Neeme Järvi', 4, 1),
 ('Herbert Von Karajan, Mirella Freni & Wiener Philharmoniker', 4, 1),
 ('Hilary Hahn, Jeffrey Kahane, Los Angeles Chamber Orchestra & Margaret Batjer', 4, 1),
 ('Itzhak Perlman', 4, 1),
 ('Leonard Bernstein & New York Philharmonic', 4, 1),
 ('Les Arts Florissants & William Christie', 4, 1),
 ('London Symphony Orchestra & Sir Charles Mackerras', 4, 1),
 ('Michele Campanella', 4, 1),
 ('Nash Ensemble', 4, 1),
 ('Orchestra of The Age of Enlightenment', 4, 1),
 ('Otto Klemperer & Philharmonia Orchestra', 4, 1),
 ('Roger Norrington, London Classical Players', 4, 1),
 ('Royal Philharmonic Orchestra & Sir Thomas Beecham', 4, 1),
 ('Sergei Prokofiev & Yuri Temirkanov', 4, 1),
 ('The 12 Cellists of The Berlin Philharmonic', 4, 1),
 ('Yehudi Menuhin', 4, 1),
 ('U2', 3, 135),
 ('Led Zeppelin', 3, 114),
 ('Deep Purple', 

### Tracks purchased vs non purchased

In order to do this we need to now: 
* Number of total tracks in the app --> total_tracks
* Number of distinct tracks in invoice_line --> tracks_purchased

In [82]:
%%sql

SELECT (SELECT count(distinct track_id) FROM invoice_line) tracks_purchased,
       count(distinct track_id) - (SELECT count(distinct track_id) FROM invoice_line) tracks_non_purchaded,
       ROUND((SELECT count(distinct track_id) FROM invoice_line)/CAST(count(distinct track_id) AS FLOAT), 3) *100 percentage,
       count(distinct track_id) total_tracks
FROM track


Done.


[(1806, 1697, 51.6, 3503)]

51.6% of the tracks available have been sold.

### Protected vs non-protected media types. Effect on sales. 

* Tenemos que crearnos una tabla donde estén todas las tracks compradas diciendo si son protected o non-protected. Para ello clasificamos los 5 tipos de media_type_id
 1,4,5 - NO PROTECTED
 2,3 - PROTECTED 
 
* A partir de ahí solo tenemos que sumar las ventas cada parte aplicando un WHERE para distinguir las dos posibilidades y sacar los porcentajes

In [83]:
%%sql 

WITH protected_track_sales AS
     (
         SELECT COUNT(il.track_id) protected_tracks_bought,
                SUM(il.unit_price) protected_sales
        FROM invoice_line il 
        INNER JOIN track t on t.track_id = il.track_id
        INNER JOIN media_type mt on mt.media_type_id = t.media_type_id
        WHERE mt.media_type_id = 2 
              OR mt.media_type_id = 3 

        ORDER BY 1 ASC
         
     )
    
SELECT (SELECT protected_tracks_bought FROM protected_track_sales) protected_tracks_sold,
       (SELECT protected_sales FROM protected_track_sales) protected_sales,
        SUM(unit_price) total_sales,
       (SELECT protected_sales FROM protected_track_sales)/SUM(unit_price) * 100 percentage
    
 FROM invoice_line    
 
       

Done.


[(442, 437.58000000000254, 4709.429999999431, 9.291570317428127)]

Protected sales are not really popular, representing roughly 9.30% of total sales. 